libraries/AP_Proximity/AP_Proximity_Boundary_3D.cpp

```diff
@@ -0,0 +1,287 @@
+#include "AP_Proximity_Backend.h"
+#include "AP_Proximity_Boundary_3D.h"
+
+/*
+  Constructor. 
+  This incorporates initialisation as well.
+*/
+AP_Proximity_Boundary_3D::AP_Proximity_Boundary_3D() 
+{
+    // initialise sector edge vector used for building the boundary fence
+    init();
+}
+
+// initialise the boundary and sector_edge_vector array used for object avoidance
+//   should be called if the sector_middle_deg or _sector_width_deg arrays are changed
+void AP_Proximity_Boundary_3D::init()
+{
+    for (uint8_t layer=0; layer < PROXIMITY_NUM_LAYERS; layer++) {
+        const float pitch = ((float)_pitch_middle_deg[layer]);
+        for (uint8_t sector=0; sector < PROXIMITY_NUM_SECTORS; sector++) {
+            const float angle_rad = ((float)_sector_middle_deg[sector]+(PROXIMITY_SECTOR_WIDTH_DEG/2.0f));
+            _sector_edge_vector[layer][sector].offset_bearing(angle_rad, pitch, 100.0f);
+            _boundary_points[layer][sector] = _sector_edge_vector[layer][sector] * PROXIMITY_BOUNDARY_DIST_DEFAULT;
+        }
+    }
+}
+
+// returns face corresponding to the provided yaw and (optionally) pitch
+// pitch is the vertical body-frame angle (in degrees) to the obstacle (0=directly ahead, 90 is above the vehicle)
+// yaw is the horizontal body-frame angle (in degrees) to the obstacle (0=directly ahead of the vehicle, 90 is to the right of the vehicle)
+AP_Proximity_Boundary_3D::Face AP_Proximity_Boundary_3D::get_face(float pitch, float yaw) const
+{   
+    const uint8_t sector = wrap_360(yaw + (PROXIMITY_SECTOR_WIDTH_DEG * 0.5f)) / 45.0f;
+    const float pitch_limited = constrain_float(pitch, -75.0f, 74.9f);
+    const uint8_t layer = (pitch_limited + 75.0f)/PROXIMITY_PITCH_WIDTH_DEG;
+    return Face(layer, sector);
+}
+
+// Set the actual body-frame angle(yaw), pitch, and distance of the detected object.
+// This method will also mark the sector and layer to be "valid", so this distance can be used for Obstacle Avoidance
+void AP_Proximity_Boundary_3D::set_face_attributes(const Face &face, float pitch, float angle, float distance)
+{
+    if (!face.valid()) {
+        return;
+    }
+
+    _angle[face.layer][face.sector] = angle;
+    _pitch[face.layer][face.sector] = pitch;
+    _distance[face.layer][face.sector] = distance;
+    _distance_valid[face.layer][face.sector] = true;
+
+    // update boundary used for simple avoidance
+    update_boundary(face);
+}
+
+// add a distance to the boundary if it is shorter than any other provided distance since the last time the boundary was reset
+// pitch and yaw are in degrees, distance is in meters
+void AP_Proximity_Boundary_3D::add_distance(float pitch, float yaw, float distance)
+{
+    Face face = get_face(pitch, yaw);
+    if (!_distance_valid[face.layer][face.sector] || (distance < _distance[face.layer][face.sector])) {
```

Comment:
You haven't checked the face is valid here.

Perhaps it should be

```
bool get_face(Face &face, pitch, yaw) WARN_IF_UNUSED
```

That would allow us to remove the validity checks that are sprinkled around for `face` - you can never hold an invalid face with this signature, so you can't pass it to functions like this one.

Note that `convert_obstacle_num_to_face` already has this interface, so ATM we're mixing techniques.

Reply:
This method can never produce a invalid face. Its been constructed so that when the user passes pitch, yaw, it ALWAYS produces a valid face. Thats why we don't have a validity check here.

Reply:
I have added a check anyway